frontend/src/metabase-lib/native.ts

```diff
@@ -1,9 +1,56 @@
 import * as ML from "cljs/metabase.lib.js";
 
-import type { DatabaseId } from "metabase-types/api";
+import type { DatabaseId, TemplateTags } from "metabase-types/api";
 
 import type { MetadataProvider, Query } from "./types";
 
+export function nativeQuery(
+  databaseId: DatabaseId,
+  metadata: MetadataProvider,
+  innerQuery: string,
+): Query {
+  return ML.native_query(databaseId, metadata, innerQuery);
+}
+
+export function rawNativeQuery(query: Query): string {
+  return ML.raw_native_query(query);
+}
+
+export function withNativeQuery(query: Query, innerQuery: string): Query {
+  return ML.with_native_query(query, innerQuery);
+}
+
+export function withTemplateTags(query: Query, tags: TemplateTags): Query {
+  return ML.with_template_tags(query, tags);
+}
+
+export function templateTags(query: Query): TemplateTags {
+  return ML.template_tags(query);
+}
+
+export function extractTemplateTags(
+  queryText: string,
+  existingTags?: TemplateTags,
+): TemplateTags | Record<string, never> {
```

Comment:
The docstring says:
> Invalid patterns are simply ignored, so something like `{{&foo!}}` is just disregarded.

So I initially (erroneously) thought this falls back to `null`. But playing around in REPL confirmed that this function returns just an empty object (map in Clojure) when it encounters an invalid pattern.

Reply:
I think in that case I'd leave only `TemplateTags` as a type because an empty object satisfies it.

Reply:
Works for me.
Fixed in https://github.com/metabase/metabase/pull/37518/commits/9ab6688f07a972c3d3ca3f51863bfc10a21ec96d

Thank you for the thorough review.